src/metabase/actions/models.clj

```diff
@@ -330,22 +332,122 @@
+     :id          (primitive-action-id kind (:id table))
+     ;; true for all databases with data_editing_enabled
+     ;; it is expected this fn will not be called if this is not the case
+     :database_enabled_actions true
+     :visualization_settings
+     {:name ""
+      :type "button"
+      :description ""
+      :fields
+      (->> (for [field field-params
+                 :let [field-name (:name field)]]
+             [field-name
+              {:hidden      false
+               :id          field-name}])
+           (into {}))}
+     :parameters
+     (->> (for [field field-params
+                :let [field-name (:name field)]]
+            {:id    field-name
+             :display-name (:display_name field)
+             :type (:base_type field)
+             :target [:variable [:template-tag field-name]]
+             :slug  field-name
+             ;; todo
+             :required (or (= :type/PK (:semantic_type field))
+                           (:database_required field))
+             :is-auto-increment (:database_is_auto_increment field)})
+          vec)}))
+
+(defn select-primitive-action
+  "Returns the primitive action (map) for the given kind (operation, e.g :row/create) and table-id."
+  [table-id kind]
+  (let [table      (t2/select-one :model/Table table-id)
+        fields     (t2/select :model/Field :table_id table-id)]
+    (table-primitive-action table fields kind)))
 
 (defn dashcard->action
   "Get the action associated with a dashcard if exists, return `nil` otherwise."
   [dashcard-or-dashcard-id]
-  (some->> (t2/select-one-fn :action_id :model/DashboardCard :id (u/the-id dashcard-or-dashcard-id))
-           (select-action :id)))
+  (let [{:keys [action_id
+                visualization_settings]}
+        (t2/select-one [:model/DashboardCard :action_id :visualization_settings] (u/the-id dashcard-or-dashcard-id))
+        {:keys [table_action]}
+        visualization_settings]
+    (cond
+      action_id (select-action :id action_id)
+      table_action
+      (let [{:keys [table_id kind]} table_action]
+        (select-primitive-action table_id (keyword kind))))))
+
+(methodical/defmethod t2.hydrate/batched-hydrate [:model/DashboardCard :dashcard/action]
+  "Hydrates actions from DashboardCards. Adds a boolean field `:database-enabled-actions` to each action according to
+  the\n `database-enable-actions` setting for the action's database."
+  [_model _k dashcards]
+  (let [actions-by-id
+        (when-let [action-ids (seq (keep :action_id dashcards))]
+          (->> (select-actions nil :id [:in action-ids])
+               map-assoc-database-enable-actions
+               (m/index-by :id)))
+        table-ids
+        (->> dashcards
+             (keep (comp :table_id :table_action :visualization_settings))
+             distinct)
+        table-by-id
+        (when (seq table-ids)
+          (->> (t2/select :model/Table :id [:in table-ids])
+               (u/index-by :id)))
```

Comment:
A bit cutesy, but you can do `(t2/select-fn->fn :id identity :model/Table ...)` to save the second indexing step.

Reply:
9baea0481f9c